src/java.base/share/classes/java/lang/VirtualThread.java

```diff
@@ -890,7 +882,19 @@ private void unblock() {
     }
 
     /**
-     * Invoked by timer thread when wait timeout for virtual thread has expired.
+     * Invoked by FJP worker thread or STPE thread when park timeout expires.
+     */
+    private void parkTimeoutExpired() {
+        assert !VirtualThread.currentThread().isVirtual();
+        if (!getAndSetParkPermit(true)
```

Comment:
@AlanBateman Would it make sense to test whether the park-permit is false before the LOCK XCHG?

Reply:
It already does, no CAS if the current value is the new value.

Reply:
I meant the park permit. :)

Reply:
getAndSetParkPermit isn't changed in this PR. If the park permit is already granted then getAndSetParkPermit(true) doesn't do anything. More generally, parkTimeoutExpired is just a reduced unpark, only added so that we can do a lazy rather than internal submit for the less common case that the timeout expires.